### PR TITLE
Rudimentary support for disassembling rb_iseq_t

### PR DESCRIPTION
**Pull request by tenderlove**
I need to disassemble instruction sequences while debugging, so I wrote
this.

Usage is like this:

```
(lldb) p iseq
(rb_iseq_t *) $147 = 0x0000000101068400
(lldb) rbdisasm iseq
0000 putspecialobject( 3 )
0002 putnil
0003 defineclass( ID: 0x560b, (rb_iseq_t *)0x1010681d0, 2 )
0007 pop
0008 putspecialobject( 3 )
0010 putnil
0011 defineclass( ID: 0x56eb, (rb_iseq_t *)0x101063b58, 2 )
0015 leave
```

### CODE REVIEW

**Comment by imomaliev** [comment]
Would it be ok to review this and suggest some small improvements from "pythonic" point of view. Or is this just some throwaway script that wouldn't be staying in code base for long?

**Comment by tenderlove** [comment]
> Would it be ok to review this and suggest some small improvements from "pythonic" point of view. Or is this just some throwaway script that wouldn't be staying in code base for long?

Hopefully we’ll be keeping this for a while since it’s for debugging with lldb.  I am not very good at Python, so I’d really appreciate some feedback so I can improve. 🙇🏻‍♂️

File: misc/lldb_disasm.py [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+TS_OFFSET   = b'O'[0]
+TS_VALUE    = b'V'[0]
+TS_LINDEX   = b'L'[0]
+TS_FUNCPTR  = b'F'[0]
+TS_NUM      = b'N'[0]
+TS_BUILTIN  = b'R'[0]
+
+def rb_vm_insn_addr2insn2(target, result, wanted_addr, size):
+    tIntPtr = target.FindFirstType("intptr_t")
+    sizeOfIntPtr = tIntPtr.GetByteSize();
```
Comment (imomaliev) [inline misc/lldb_disasm.py:41]: `;` could be deleted

File: misc/lldb_disasm.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+    addr_of_table = target.FindSymbols("vm_exec_core.insns_address_table")[0].GetSymbol().GetStartAddress().GetLoadAddress(target)
+
+    for insn in range(0, size):
```
Comment (imomaliev) [inline misc/lldb_disasm.py:44]: `range(0, size)` -> could be `range(size)`, `0` redundant here

https://docs.python.org/3.8/library/functions.html#func-range

File: misc/lldb_disasm.py [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+    print("%04d %s" % (n, name), file=result, end="")
+
+    if length == 1:
+        print("", file=result)
+        return length
+
+    idx = 0
+    print("(", end="", file=result)
+    for op_type in op_types:
```
Comment (imomaliev) [inline misc/lldb_disasm.py:74]: `for op_type in op_types:` -> `for idx, op_type in enumerate(op_types):` this way you wouldn't have to manually increment `idx`.

https://docs.python.org/3.8/library/functions.html#enumerate

File: misc/lldb_disasm.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+            print("UNREACHABLE?" % op, end="", file=result)
+
+        idx += 1
+
+    print(" )", file=result)
+    return length
+
+def insn_len(target, offset):
+    tChar = target.FindFirstType("char")
+    size_of_char = tChar.GetByteSize();
```
Comment (imomaliev) [inline misc/lldb_disasm.py:122]: `;` could be deleted

File: misc/lldb_disasm.py [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+    addr_in_table = addr_of_table + (offset * size_of_char)
+    addr = lldb.SBAddress(addr_in_table, target)
+
+    return target.CreateValueFromAddress("y", addr, tChar).GetValueAsUnsigned()
+
+def insn_op_types(target, process, result, insn):
+    tUShort = target.FindFirstType("unsigned short")
+    tChar   = target.FindFirstType("char")
+
+    size_of_short = tUShort.GetByteSize();
```
Comment (imomaliev) [inline misc/lldb_disasm.py:135]: `;` could be deleted

File: misc/lldb_disasm.py [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+    size_of_char =  tChar.GetByteSize();
```
Comment (imomaliev) [inline misc/lldb_disasm.py:136]: `;` could be deleted

File: misc/lldb_disasm.py [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+
+    addr_of_table = target.FindSymbols("insn_op_types.x")[0].GetSymbol().GetStartAddress().GetLoadAddress(target)
+    addr_in_name_table = addr_of_table + (offset * size_of_char)
+
+    error = lldb.SBError()
+    return process.ReadCStringFromMemory(addr_in_name_table, 256, error)
+
+def insn_name_table_offset(target, offset):
+    tUShort = target.FindFirstType("unsigned short")
+    size_of_short = tUShort.GetByteSize();
```
Comment (imomaliev) [inline misc/lldb_disasm.py:151]: `;` could be deleted

**Review by imomaliev — CHANGES_REQUESTED** [review]
Here are small improvements. I will have more time on monday to properly look into this

File: misc/lldb_disasm.py (hunk 7 above)
Comment (tenderlove) [reply misc/lldb_disasm.py:151]: argh!  Thanks for pointing these out. I guess I was flipping between C and Python too much 🤦🏻‍♂️

File: misc/lldb_disasm.py [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+        if idx == 0:
+            print(" ", end="", file=result)
+        else:
+            print(", ", end="", file=result)
+
+        opAddr = lldb.SBAddress(iseqs.unsigned + ((n + idx + 1) * sizeofValueP), target)
+        opValue = target.CreateValueFromAddress("op", opAddr, tValueP)
+        op = opValue.GetValueAsUnsigned()
+
+        if op_type == TS_VARIABLE:
```
Comment (imomaliev) [inline misc/lldb_disasm.py:83]: What do you think about refactoring this out to use `dict` with `op_type` as keys, instead of this big `if` statement?

Comment (tenderlove) [reply misc/lldb_disasm.py]: done!

**Comment by kivikakk** [comment]
🎉

**Comment by imomaliev** [comment]
Does this need to be python 2 compatible?

**Pull request merged**